kubernetes/cluster_deploy.go

```diff
@@ -14,38 +15,77 @@ import (
 )
 
 func (c *Cluster) Deploy(options *ltops.DeployOptions) error {
-	log.Info("installing mattermost helm chart...")
-
-	if len(c.ReleaseName) > 0 {
-		log.Info("already installed as release '" + c.ReleaseName + "'")
-		return nil
+	previouslyDeployed := len(c.Release()) > 0
+	if previouslyDeployed {
+		log.Info("upgrading mattermost helm chart...")
+	} else {
+		log.Info("installing mattermost helm chart...")
 	}
 
 	license, err := ltops.GetFileOrURL(options.LicenseFile)
 	if err != nil {
 		return err
 	}
 
-	config, err := c.GetHelmConfigFromProfile(options.Profile, options.Users, string(license))
-	if err != nil {
-		return err
-	}
+	configFileLocation := ""
 
-	err = saveChartConfig(config, c.Config.WorkingDirectory)
-	if err != nil {
-		return err
+	if len(options.HelmConfigFile) == 0 {
+		config, err := c.GetHelmConfigFromProfile(options.Profile, options.Users, string(license))
+		if err != nil {
+			return err
+		}
+
+		err = saveChartConfig(config, c.Config.WorkingDirectory)
+		if err != nil {
+			return err
+		}
+
+		configFileLocation = filepath.Join(c.Config.WorkingDirectory, chartFilename)
+	} else {
+		configFileLocation = options.HelmConfigFile
 	}
 
-	cmd := exec.Command("helm", "install", "-f", filepath.Join(c.Config.WorkingDirectory, chartFilename), "mattermost/mattermost-helm")
-	out, err := cmd.CombinedOutput()
-	if err != nil {
-		return errors.Wrap(err, "unable to install mattermost chart, error from helm: "+string(out))
+	if previouslyDeployed {
+		cmd := exec.Command("helm", "upgrade", "-f", configFileLocation, c.Release(), "mattermost/mattermost-helm")
+		if out, err := cmd.CombinedOutput(); err != nil {
+			return errors.Wrap(err, "unable to install mattermost chart, error from helm: "+string(out))
+		}
+
+		// Delete the pods so they are recreated with any config changes
+		cmd = exec.Command("kubectl", "delete", "po", "-l", fmt.Sprintf("release=%v", c.Release()))
+		if out, err := cmd.CombinedOutput(); err != nil {
+			return errors.Wrap(err, "unable to restart release pods, error from kubectl: "+string(out))
+		}
+
+		cmd = exec.Command("kubectl", "delete", "po", "-l", fmt.Sprintf("app=%v-mysqlha", c.Release()))
+		if out, err := cmd.CombinedOutput(); err != nil {
+			return errors.Wrap(err, "unable to restart mysqlha pods, error from kubectl: "+string(out))
+		}
+
+		log.Info("upgraded release '" + c.ReleaseName + "'")
+	} else {
+		cmd := exec.Command("helm", "install", "-f", configFileLocation, "mattermost/mattermost-helm")
```

Comment:
Out of scope for this PR, but given that kubernetes is written in Go, is there an API for effecting these and getting more discrete output vs. just executing the equivalent `helm`/`kubectl` commands?

Reply:
There is yes, the kubectl one is actually already imported and used here anywhere I need to get some information and parse it. I also started using the helm Go library as well but I ran into some difficulties with dependencies and I was wasting a lot of mana on it so I just switched to shelling out.

I would like to stop shelling out as much as possible but I've mainly been going for whatever is fastest and works